### PR TITLE
Store extended colors in v:colornames dict.

### PR DESCRIPTION
**Pull request by dvogel**
Problem solved: Remembering RGB colors is difficult.
Solution: Allow user to (re-)define new color names.

---

This is a follow-up to a previous pull request (https://github.com/vim/vim/pull/8426). The aim is the same but
this implementation tries to follow @brammool's suggestion to use a builtin dictionary to make manipulation
and examination of the color table more natural.

One point of friction with this approach relates to the casing of color names. The names loaded from rgb.txt
are converted to lowercase when they are read into the table. The lookups are also done in lowercase. This
satisfies all of the existing mixed-case uses of the color names. However, it could be slightly confusing for
users accustomed to the casing found in rgb.txt. It should be apparent upon examination of the dictionary,
such as with `let v:colornames`.

The color lookup will accommodate numeric values in `v:colornames` but the expectation is to use hex color
strings. The hex strings only require decoding when used in conjunction with a `highlight` command, so the
cost is minimal and the usability improvement is substantial.

### CODE REVIEW

**Comment by brammool** [comment]
I'll leave this open for others to give comments.

The code can probably go in src/highlight.c, instead of adding a new colornames.c file.

File: runtime/doc/eval.txt [hunk 1]
```diff
@@ -1830,6 +1830,22 @@ v:collate	The current locale setting for collation order of the runtime
 		command.
 		See |multi-lang|.
 
+                                                                *v:colornames*
+v:colornames    A dictionary that maps color names to hex color strings. These
+                color names can be used with the |highlight-guifg| and
+		|highlight-guibg| parameters. The dictionary is provided only
```
Comment (romainl) [inline runtime/doc/eval.txt]: `|highlight-guisp|` is missing.

**Comment by romainl** [comment]
What would be the expected outcome when the user does something like:

    :let v:colornames = {}

or something of the sort? I understand that it wouldn't change anything for a colorscheme using those special names that is already sourced but what if you remove all those names and source that colorscheme again? I guess we are expecting that list to be an explicit part of colorschemes, a bit like with `g:terminal_ansi_colors`, right?

Another question: suppose users source two colorschemes that use that mechanism, shouldn't we devise something that clears `g:colornames` before defining it with our own palette? Or some "merging"strategy? In that case, what would happen to the default values that could be used in `default`, for example, or highlight groups forgotten by the author?

**Comment by dvogel** [comment]
> What would be the expected outcome when the user does something like:
> 
> ```
> :let v:colornames = {}
> ```
> 
> or something of the sort?

Thanks for the feedback @romainl. Since `v:colornames` is a read-only variable (in the vim var sense) the user will get an error if they try to assign a new dict to it. The elements are over-writable and can be removed with `unlet`. I've just pushed a new commit that adds these to `src/testdir/test_colornames.vim`.

> what if you remove all those names and source that colorscheme again? I guess we are expecting that list to be an explicit part of colorschemes, a bit like with `g:terminal_ansi_colors`, right?

...

> Another question: suppose users source two colorschemes that use that mechanism, shouldn't we devise something that clears `g:colornames` before defining it with our own palette? Or some "merging"strategy? In that case, what would happen to the default values that could be used in `default`, for example, or highlight groups forgotten by the author?

The last script to write before a `highlight` command will "win". There's two ideas here:

* The idea is for colorschemes that want a very specific color to use a very specific color name and include the color definition in the colorscheme file. I've [been using this](https://github.com/dvogel/dotfiles/blob/vim-namecolor/vim/.vim/colors/drew.vim#L18) with my color names prefixed with "drew_" to avoid collisions.
* Conversely, colorschemes that want to rely on common color names can avoid defining the colors in the colorscheme file, and rely on vim runtime files (such as `runtime/plugins/csscolors.vim`) to define those colors. In this case colorscheme scripts would either need to source the common color definition scripts or the user would need to trigger their colorscheme after the common colors have been loaded (e.g. by creating an `after/plugin/csscolors.vim`).

In theory a script could clear the dictionary using normal vimscript:

```
for k in keys(v:colornames)
    unlet v:colornames[k]
endfor
```

This would cause havoc for the rgb.txt colors since those are only loaded once. Outside of debugging, I'm not sure of a reason to clear the dictionary though since any script that might clear it would be able to overwrite an existing entry the script wanted to control. We could easily make the code re-load the rgb.txt colors anytime one is not found if we foresee good reasons people might clear the dictionary.

**Comment by dvogel** [comment]
> I'll leave this open for others to give comments.
> 
> The code can probably go in src/highlight.c, instead of adding a new colornames.c file.

Sounds good. I will move the `colornames.c` code into `hightlight.c` once I incorporate the other feedback, just to keep churn low.

**Comment by chrisbra** [comment]
> The last script to write before a highlight command will "win". There's two ideas here:

So what happens, with an existing highlighting command, when I change the color definition later?

> for k in keys(v:colornames)
>     unlet v:colornames[k]
> endfor

Hm, does it make sense to check, that the dictionary contains at least xxx entries? Or that at least the default colors (whatever they are...) are always defined?

File: src/colornames.c [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+    dictitem_T *existing;
+    char_u     hexstr[8];
+
+    if (snprintf((char *)hexstr, sizeof(hexstr), "#%02x%02x%02x", r, g, b) < 0)
+    {
+	semsg(_(e_alloc_color), name);
+	return;
+    }
+
+    dict_T *colornames_table = get_vim_var_dict(VV_COLORNAMES);
```
Comment (chrisbra) [inline src/colornames.c]: you need to move the declaration of variables to the beginning of the block.

Comment (dvogel) [reply src/colornames.c]: Thanks for catching those!

**Comment by dvogel** [comment]
> > The last script to write before a highlight command will "win". There's two ideas here:
> 
> So what happens, with an existing highlighting command, when I change the color definition later?

Nothing changes at that time. The only time the highlight table is updated is when the `highlight` command is executed. The `v:colornames` dict is only for allowing system-defined colors and user-defined colors to be referenced in the same way.

> Hm, does it make sense to check, that the dictionary contains at least xxx entries? Or that at least the default colors (whatever they are...) are always defined?

Personally, I don't think so. That would be one way to retain the current behavior (rgb.txt colors cannot be unloaded once they are loaded in the current implementation). However, I find the new malleability of those colors to be a benefit. I hope to be able to more easily override the definition of those colors (my eyes benefit from higher saturation). It would take considerable intent to decide to clear the `v:colornames` dict without intending to override default behaviors.

**Comment by brammool** [comment]
I think what would be good to have is a section in the help where the
use of v:colornames is explained, with examples.

This should cover both personal use, e.g. setting v:colornames in your
.vimrc or another file under ~/.vim, and use with plugins and syntax
files for various filetypes.

Once we have that we can decide if just having one dictionary for the
whole of Vim, which can be changed any time, is a good solution.
It might be that we need a more complicated solution, e.g. a list of
dictionaries, so that plugins can add and remove their own set.

-- 
I AM THANKFUL...
...for the clothes that fit a little too snug because it
means I have more than enough to eat.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Pull request closed without merging**